### DISCUSSION TITLE
Android Jobs for calendar checking

### DISCUSSION THREAD

**Discussion by pabloat81** [shinyorg/shiny#1002] [Q&A] [answered]
Hello.
I need to check every 15 mins the database to check if is there any calendar (my app has a calendar module) event in the next minutes and raise a notification.

I need this to run always, also with the App not running. 
It is possible with jobs?

Also, I am copying from the jobs/sample project but something i missing. 
I have Shiny, Shiny.core, Shiny.jobs and Shiny.notifications (on shared and android project).
Added code in assemblyinfo.cs, attributes on main activity, created startup.cs, samplejob.cs, jobloggettask.cs
But nothig gets called, even no breakpoint reached.

I am missing something? Any dependency?

I didnt put any "Init" method in main activity because code sample doesnt has any of that.

Thanks in advance.

**Comment by pabloat81** [comment]
I forgot to mention i have a SplashScreen activity, i tried putting:
[assembly: Shiny.ShinyApplication(
    ShinyStartupTypeName = "xxxxx.Startup",
    XamarinFormsAppTypeName = "xxxxx.App"
)]
in splashscreen and in mainactivity

**Reply by aritchie** [reply]
> I need to check every 15 mins the database to check if is there any calendar (my app has a calendar module) event in the next minutes and raise a notification.

This is stated all over the samples and docs - jobs are not scheduled, they are periodic.  The OS decides if network, battery, etc amongst other things is in a state that jobs should run.  Don't assume scheduled times ever.  With notifications, you can use time sensitive/scheduled notifications that are part of v2.5.


> But nothig gets called, even no breakpoint reached.
> I am missing something? Any dependency?

I don't have enough information to make a guess.  You have to provide a SMALL reproducible sample.  I can tell you

* You can't debug a job in the background with just "sit and wait".  There are tricks you can do with adb on Android.  There are also other explanations I have given on how to test/debug jobs on iOS using Vs4mac - search it up.   
* There are rumblings of issues with splash & main activity using the code gen.  No one has ever sent a reproducible sample for me to verify with.  In most cases, the code gen is only there to get you started, you should hardwire the shiny calls if you are doing anything differently  and only wire things up manually to ONLY your main activity.

**Reply by pabloat81** [reply]
Hi Allan, thanks for the response (sorry my english)

I am new to your lib, i saw some videos on YT and i was looking at the docs pages at https://shinylib.net/quickstart (for some days is giving HTTP ERROR 404, so i was trying with the samples because i cannot find DOCs, maybe where moved to other address?).


I understand that jobs depends on many factors, but for the little i saw till now i thougth that for what i need jobs are the  best option. Do you say there is a better approach? I need to notify my users for next events only, an then maybe when click notification redirects to my app on that event.

I have to get from nuget, Shiny, Shiny.Core, Shiny.Notifications and Shiny.Jobs? in shared and Android project?

I didnt understand the:
"you can use time sensitive/scheduled notifications that are part of v2.5." because i cant find the Docs.

Also for the:
"here are rumblings of issues with splash & main activity using the code gen."
and "you should hardwire the shiny calls if you are doing anything differently and only wire things up manually to ONLY your main activity."
How do i "hardwire", with a Init?
Also, the codegen should generate some code? when running the project nothing gets generated.

Last, for:
"You have to provide a SMALL reproducible sample"
I could copy some snnipets of all i done?

Thanks!

**Comment by aritchie** [comment, marked as answer]
https://shinylib.net is where the docs are.  github.com/shinyorg/samples are where are the samples are.

**Reply by pabloat81** [reply]
Thanks Allan.
I readed all the doc several times and made some missing changes. I think i made all steps.

I have a SplashActivity (: AppCompatActivity), maybe this is still my problem?

I tried Shiny.ShinyApplication attribute on Splash and mainactivity.

In Shared app, on dependency->analyzers now appears a Shiny.generators (empty)
In Android project dont appears anything in analyzers.

Any posible sugestion?

Thanks.

**Reply by aritchie** [reply]
I can’t guess.  You have to supply a small reproducible sample.  The samples show many with and without the source generators

**Reply by pabloat81** [reply]
Thanks Allan,
I made it work without the source generators.

Two questions, the docs page are a bit simple:
Is it possible to maintain the job running even with the app killed or on a restart?

On a notification tap, its possible to pass some parameter to point to the correct agenda event?

Thanks.

**Comment by aritchie** [comment]
> Two questions, the docs page are a bit simple
Remember that this is not a product.  You are always welcome to contribute back :)

> Is it possible to maintain the job running even with the app killed or on a restart?
Jobs only run for a small amount of time.  If the app is killed or the system is rebooted, they will pick up later.  Again, the timing is decided mostly by the OS

> On a notification tap, its possible to pass some parameter to point to the correct agenda event?
Yes, you can read the notification.Payload

**Reply by pabloat81** [reply]
Ok, thanks.

I couldnt make it work killing the app on emulator nor my phone, maybe some permission or constructor property?

**Reply by pabloat81** [reply]
I am trying with the snippet below (tried some changes on the options too).
I setted 20 secs for testing, but with the app in foreground excecutes ramdomly maybe after some seconds o maybe minutes sometimes (i know this is normal). But with the app in backgrownd or killing it i cant get it excecuted.

`public override void ConfigureServices(IServiceCollection services, IPlatform platform)
        {
            services.AddSingleton<UnitOfWork>();
            services.AddSingleton<JobLoggerTask>();
            services.UseJobs(true);
            services.UseNotifications(); 
            services.RegisterJob(new JobInfo(typeof(CalendarJob))
            {
                //RunOnForeground = true,
                RunOnForeground = false,
                DeviceCharging = false,
                BatteryNotLow = false,
                Repeat = true,
                RequiredInternetAccess = InternetAccess.None,
                PeriodicTime = new TimeSpan(0, 0, 30)                
            });
        }`

**Reply by aritchie** [reply]
Please read on the Android workmanager.  This isn't meant as a general tech support forum.  The job will get executed in the background but it can take anywhere from 15mins to a couple of hours depending on the doze settings.
